pkg/cmd/cli/cmd/observe/observe.go

```diff
@@ -0,0 +1,936 @@
+package observe
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+	"text/template"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/healthz"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/jsonpath"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/util/proc"
+)
+
+var (
+	observeCounts = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "observe_counts",
+			Help: "Number of changes observed to the underlying resource.",
+		},
+		[]string{"type"},
+	)
+	execDurations = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "observe_exec_durations_milliseconds",
+			Help: "Item execution latency distributions.",
+		},
+		[]string{"type", "exit_code"},
+	)
+	nameExecDurations = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "observe_name_exec_durations_milliseconds",
+			Help: "Name list execution latency distributions.",
+		},
+		[]string{"exit_code"},
+	)
+)
+
+const (
+	observeLong = `
+Observe changes to resources and take action on them
+
+This command assists in building scripted reactions to changes that occur in
+Kubernetes or OpenShift resources. These reactions are applied to every object
+periodically and are known as 'controllers'.
+
+The simplest pattern is maintaining an invariant on an object - for instance,
+"every namespace should have an annotation that indicates its owner". If the
+object is deleted no reaction is necessary. A variation on that pattern is
+creating another object: "every namespace should have a quota object based
+on the resources allowed for an owner".
+
+		$ cat set_owner.sh
+		#!/bin/sh
+		if [[ "$(%[1]s get namespace "$1" --template='{{ .metadata.annotations.owner }}')" == "" ]]; then
+			%[1]s annotate namespace "$1" owner=bob
+		fi
+
+    $ %[1]s observe namespaces --each ./set_owner.sh
+
+The set_owner.sh script is invoked with a single argument (the namespace name)
```

Comment:
Seems like every namespaced resource is going to need to be called with a `name` as well.